test/files/admin_role_custom.lua

```diff
@@ -0,0 +1,143 @@
+
+        if opts.age ~= nil then
+            table.insert(res, string.format("You are %s years old", opts.age))
+        else
+            table.insert(res, string.format("I don't know your age"))
+        end
+
+        if opts.loves_cakes then
+            table.insert(res, string.format("I know that you like cakes!"))
+        else
+            table.insert(res, string.format("How can you not love cakes?"))
+        end
+
+        return res
+    end,
+}
+
+local func_long_name = {
+    usage = 'func_long_name usage',
+    call = function() return {string.format('func_long_name was called')} end,
+}
+
+local func_no_args = {
+    usage = 'func_no_args usage',
+    call = function() return {string.format('func_no_args was called')} end,
+}
+
+local func_long_arg = {
+    usage = 'func_long_arg usage',
+    args = {
+        long_arg = {
+            usage = 'long_arg usage',
+            type = 'string',
+        }
+    },
+    call = function(opts)
+        return {
+            string.format('func_long_arg was called with %q arg', opts.long_arg),
+        }
+    end,
+}
+
+local func_bad_arg_type = {
+    usage = 'func_bad_arg_type usage',
+    args = {
+        bad_arg = {
+            usage = 'bad_arg usage',
+            type = 'bad-type',
+        }
+    },
+    call = function() return {string.format('func_long_arg was called')} end,
+}
+
+local func_rets_str = {
+    usage = 'func_rets_str usage',
+    call = function() return string.format('func_rets_str was called') end,
+}
+
+local func_rets_non_str = {
+    usage = 'func_rets_non_str usage',
+    call = function() return 666 end,
+}
+
+local func_conflicting = {
+    usage = 'func_conflicting usage',
+    args = {},
+    call = function() return {
+        string.format('func_conflicting shouldn\'t be called'),
+    } end,
+}
+
+local conflicting_names = {
+    "name", "instance", "run_dir", "data_dir", "list", "help",
+    "debug", "quiet", "verbose",
+}
+
+for _, argname in ipairs(conflicting_names) do
+    func_conflicting.args[argname] = {
+        usage = string.format('%s usage', argname),
+        type = 'string',
+    }
+end
+
+local function init()
+    admin.init()
+
+    admin.register('echo_user', echo_user.usage, echo_user.args, echo_user.call)
+    admin.register('func.long.name', func_long_name.usage, func_long_name.args, func_long_name.call)
+    admin.register('func_no_args', func_no_args.usage, func_no_args.args, func_no_args.call)
+    admin.register('func_long_arg', func_long_arg.usage, func_long_arg.args, func_long_arg.call)
+    admin.register('func_bad_arg_type', func_bad_arg_type.usage, func_bad_arg_type.args, func_bad_arg_type.call)
+    admin.register('func_rets_str', func_rets_str.usage, func_rets_str.args, func_rets_str.call)
+    admin.register('func_rets_non_str', func_rets_non_str.usage, func_rets_non_str.args, func_rets_non_str.call)
+    admin.register('func_conflicting', func_conflicting.usage, func_conflicting.args, func_conflicting.call)
+
```

Comment:
I don't see a case when function raises error

Reply:
Added cases when function raises and returns error.